src/main/kotlin/com/toasttab/pulseman/state/TabState.kt

```diff
@@ -0,0 +1,206 @@
+
+    private val sendMessage = SendMessage(
+        setUserFeedback = userFeedback::setUserFeedback,
+        selectedClass = protobufSelector.selectedSendClass,
+        pulsarSettings = pulsarSettings,
+        initialSettings = initialSettings,
+        onChange = ::onChange
+    )
+    private val receiveMessage = ReceiveMessage(
+        setUserFeedback = userFeedback::setUserFeedback,
+        pulsarSettings = pulsarSettings,
+        selectedReceiveClasses = protobufSelector.selectedReceiveClasses
+    )
+
+    // Tab icon settings
+    private val focused = mutableStateOf(false)
+    private val image = mutableStateOf(Icons.Default.Circle)
+    private val drawBackground = mutableStateOf(false)
+
+    private fun onEnterIconUnsavedChanges() {
+        drawBackground.value = true
+        image.value = Icons.Default.Close
+    }
+
+    private fun onExitIconUnsavedChanges() {
+        drawBackground.value = false
+        image.value = Icons.Default.Circle
+    }
+
+    private fun onEnterIcon() {
+        drawBackground.value = true
+    }
+
+    private fun onExitIcon() {
+        drawBackground.value = false
+    }
+
+    private fun onFocusedUpdate(newValue: Boolean) {
+        focused.value = newValue
+    }
+
+    private fun onTabNameChange(newValue: String) {
+        tabName.value = newValue
+        onChange()
+    }
+
+    private fun onChange() {
+        unsavedChanges.value = lastSavedTabValues != tabValues()
+    }
+
+    @ExperimentalFoundationApi
+    @ExperimentalComposeUiApi
+    @Composable
+    fun toTab(): Tab {
+        return Tab(
+            tabName = tabName.value,
+            close = { close(this) },
+            unsavedChanges = unsavedChanges.value,
+            initialSettings = initialSettings,
+            isActive = isActive,
+            onFocusedUpdate = ::onFocusedUpdate,
+            activate = ::activate,
+            image = image.value,
+            onEnterIconUnsavedChanges = ::onEnterIconUnsavedChanges,
+            onExitIconUnsavedChanges = ::onExitIconUnsavedChanges,
+            onEnterIcon = ::onEnterIcon,
+            onExitIcon = ::onExitIcon,
+            drawBackground = drawBackground.value,
+            focused = focused.value,
+            ui = {
+                tabUI(
+                    tabName = tabName.value,
+                    onTabNameChange = ::onTabNameChange,
+                    selectedView = selectedView.value,
+                    userFeedbackUI = userFeedback.ui(),
+                    messageClassSelectorUI = protobufSelector.getUI(),
+                    pulsarSettingsUI = pulsarSettings.getUI(),
+                    receiveMessageUI = receiveMessage.getUI(),
+                    sendMessageUI = sendMessage.getUI(),
+                    selectTabViewUI = {
+                        selectTabViewUI(selectedView.value, selectedView::onStateChange)
+                    }
+                )
+            }
+        )
+    }
+}
```

Comment:
This is near the top of the view hierarchy where all the ui elements are pulled in